email/fetcher.go

```diff
@@ -0,0 +1,393 @@
+package email
+
+import (
+	"abuse-scanner/database"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+
+	"github.com/emersion/go-imap"
+	"github.com/emersion/go-imap/client"
+)
+
+const (
+	// fetchFrequency defines the frequency with which we fetch new emails
+	fetchFrequency = 30 * time.Second
+)
+
+type (
+	// Fetcher is an object that will periodically scan an inbox and persist the
+	// missing messages in the database.
+	Fetcher struct {
+		staticContext     context.Context
+		staticDatabase    *database.AbuseScannerDB
+		staticEmailClient *client.Client
+		staticLogger      *logrus.Entry
+		staticMailbox     string
+	}
+
+	// messageIDMap is a helper struct that maps the message uid to its seq num
+	messageIDMap map[uint32]uint32
+)
+
+// NewFetcher creates a new fetcher.
+func NewFetcher(ctx context.Context, database *database.AbuseScannerDB, emailClient *client.Client, mailbox string, logger *logrus.Logger) *Fetcher {
+	return &Fetcher{
+		staticContext:     ctx,
+		staticDatabase:    database,
+		staticEmailClient: emailClient,
+		staticLogger:      logger.WithField("module", "Fetcher"),
+		staticMailbox:     mailbox,
+	}
+}
+
+// Start initializes the fetch process.
+func (f *Fetcher) Start() error {
+	// list mailboxes
+	mailboxes, err := f.listMailboxes()
+	if err != nil {
+		return err
+	}
+
+	// check whether the mailbox exists
+	found := false
+	for _, m := range mailboxes {
+		if m == f.staticMailbox {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return errors.New("mailbox not found")
+	}
+
+	go f.threadedFetchMessages()
+	return nil
+}
+
+// listMailboxes returns all mailboxes
+func (f *Fetcher) listMailboxes() ([]string, error) {
+	// convenience variables
+	email := f.staticEmailClient
+
+	// list all mailboxes
+	mailboxesChan := make(chan *imap.MailboxInfo, 10)
+	done := make(chan error, 1)
+	go func() {
+		done <- email.List("", "*", mailboxesChan)
+	}()
+
+	// add all inboxes to an array
+	mailboxes := make([]string, 0)
+	for m := range mailboxesChan {
+		mailboxes = append(mailboxes, m.Name)
+	}
+
+	// check whether an error occurred
+	err := <-done
+	if err != nil {
+		return nil, err
+	}
+	return mailboxes, nil
+}
+
+// threadedFetchMessages will periodically fetch new messages from the mailbox.
+func (f *Fetcher) threadedFetchMessages() {
+	// convenience variables
+	logger := f.staticLogger
+	first := true
+
+	// start the loop
+	for {
+		// sleep until next iteration, sleeping at the start of the for loop
+		// allows to 'continue' on error.
+		if !first {
+			select {
+			case <-f.staticContext.Done():
+				logger.Debugln("Fetcher context done")
+				return
+			case <-time.After(fetchFrequency):
```

Comment:
I'll take a page from Chris's book and say that a ticker would be more efficient here.